Mathlib/Data/NNRat/Defs.lean

```diff
@@ -25,6 +25,11 @@ of `x` with `↑x`. This tactic also works for a function `f : α → ℚ` with
 ## Notation
 
 `ℚ≥0` is notation for `NNRat` in locale `NNRat`.
+
+## Huge warning
+
+Whenever you state a lemma about the coercion `ℚ≥0 → ℚ`, check that Lean inserts `NNRat.cast`, not
+`Subtype.val`. Else your lemma will never apply.
```

Comment:
Ouch, this is rather unfortunate.

Reply:
Yeah... The precise issue is that `⟨q, hq⟩` notation creates a term of type `{q : ℚ // 0 ≤ q}`, not `ℚ≥0`, so it finds `Subtype.val` as the coercion, not `NNRat.cast`

Reply:
This suggests we might want an `NNRat.mk` that has the correct return type?

But if so, that can wait until a later PR

bors merge

Reply:
Maybe, but it won't have the constructor notation. To properly solve it we should either:
* make `NNRat` reducible
* make `NNRat` a one-field structure
* make `NNRat` a fully fledged structure with `num den : Nat` etc